File: 54SpiralMatrix.py
    def sizeof(self,row,col):
        return row*col
    def spiralOrder(self, matrix: List[List[int]]) -> List[int]:
        left = 0
        right = len(matrix[0])-1
        top = 0
        bottom = len(matrix)-1
        size = len(matrix) * len(matrix[0])
        res = []
        while len(res) < size:
            i = left
            while i <= right and len(res) < size:
                res.append(matrix[top][i])
                i+=1
            print(res)
            top += 1
            i = top
            while i <= bottom and len(res) < size:
                res.append(matrix[i][right])
                i+=1
            print(res)
            right -= 1
            i = right
            while i >= left and len(res) < size:
                res.append(matrix[bottom][i])
                i -= 1
            print(res)
            bottom -= 1
            i = bottom
            while i >= top and len(res) < size:
                res.append(matrix[i][left])
                i-=1
            left += 1
            print(res)
        return res
    
                

File: 54.SpiralMatrix.java
class Solution {
    public List<Integer> spiralOrder(int[][] matrix) {
        List<Integer> result = new ArrayList<>();
        
        if (matrix.length == 0){
            return result;
        }

        int left = 0;
        int right = matrix[0].length - 1;
        int top = 0;
        int bottom = matrix.length - 1;

        while (top <= bottom && left <= right){
            
            // top left -> top right
            for (int j = left; j <= right; j++){
                result.add(matrix[top][j]);
            }
            top++;
            
            // top right -> bottom right
            for (int j = top; j <= bottom; j++){
                result.add(matrix[j][right]);
            }
            right--;
            

            if (top <= bottom){
                // bottom right -> bottom left
                for (int j = right; j >= left; j--){
                    result.add(matrix[bottom][j]);
                }
            }
            
            bottom--;
            
            if (left <= right){
                // bottom left -> top left
                for( int j = bottom; j >= top; j--){
                    result.add(matrix[j][left]);
                }
            }
            left++;

        }
       
       return result;
        
    }
}
